### PR TITLE
chore: add example benchmark tool

### PR DESCRIPTION
**Pull request by dengmingtong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [ ] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

**Review by zxkane — CHANGES_REQUESTED** [review]
- [x] move a separate folder like `benchmark`
- [x] refacor to reuse existing standalone generator

File: examples/standalone-data-generator/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+requests~=2.28.2
```
Comment (zxkane) [inline examples/standalone-data-generator/requirements.txt:1]: remove this file

**Review by zxkane — CHANGES_REQUESTED** [review]

File: examples/standalone-data-generator/requirements.txt (hunk 1 above)
Comment (dengmingtong) [reply examples/standalone-data-generator/requirements.txt]: Done

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.80|< 80|
|Duplicated lines density|:white_check_mark: OK|8.60|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/12/2024, 08:24:40 (UTC+0)_

**Comment by zxkane** [comment]
# SonarQube Code Analytics
## Quality Gate passed

passed

## Additional information
*The following metrics might not affect the Quality Gate status but improving them will improve your project code quality.*

## Issues
bug   [0 Bugs](http://sonarqube:9000/project/issues?id=pr-1057&resolved=false&sinceLeakPeriod=true&types=BUG)

vulnerability   [0 Vulnerabilities](http://sonarqube:9000/project/issues?id=pr-1057&resolved=false&sinceLeakPeriod=true&types=VULNERABILITY)

code_smell   [0 Code Smells](http://sonarqube:9000/project/issues?id=pr-1057&resolved=false&sinceLeakPeriod=true&types=CODE_SMELL)

## Coverage and Duplications
*No data* **Coverage**

*No data* **Duplication**

**Comment by zxkane** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|8.60|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/12/2024, 08:28:33 (UTC+0)_

File: examples/benchmark/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# run below command on M1 mac to build image for linux/amd64
+FROM --platform=linux/amd64 python:3-alpine   
+
+# Set the working directory to /app
+WORKDIR /app
+
+# Copy the current directory contents into the container at /app
+COPY standalone-data-generator/ /app
+COPY . /app
```
Comment (zxkane) [inline examples/benchmark/Dockerfile:9]: Will `.` include `standalone-data-generator` too?

File: examples/benchmark/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+# Install any needed packages specified in requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
```
Comment (zxkane) [inline examples/benchmark/Dockerfile:12]: Do we have this file?

File: examples/benchmark/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# Make the entrypoint script executable
+RUN chmod +x /app/performance-entrypoint.sh
+
+RUN ls -la /app
+
+# Set the default command to run when the container starts
+CMD ["python", "create_event_for_performance.py"]
```
Comment (zxkane) [inline examples/benchmark/Dockerfile:20]: rename file to `create_benchmark_events`

File: examples/benchmark/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+# Copy the current directory contents into the container at /app
+COPY standalone-data-generator/ /app
+COPY . /app
+
+# Install any needed packages specified in requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
+
+# Make the entrypoint script executable
+RUN chmod +x /app/performance-entrypoint.sh
```
Comment (zxkane) [inline examples/benchmark/Dockerfile:15]: rename to `benchmark-entrypoint.sh`

File: examples/benchmark/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+## Clickstream benchmark toole
```
Comment (zxkane) [inline examples/benchmark/README.md:1]: typo `toole`

File: examples/benchmark/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+We have developed a tool for performance tests, which utilizes ECS Fargate to continuously send test data to Clickstream. The number of ECS Service tasks can be increased or decreased to adjust the RPS (Requests per Second).
```
Comment (zxkane) [inline examples/benchmark/README.md:3]: test -> mock

File: examples/benchmark/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+Follow below step to start your Clickstream performance test
+1. Add configuration file:
+
+Please download your `amplifyconfiguration.json` file from your web console of the solution, then copy it to the root
+folder which at the same level as README file. then you can execute the following command to create and send your sample
+data.
+
+In the `amplifyconfiguration.json` file, we will only parse `appId`, `endpoint` and `isCompressEvents` three parameter
```
Comment (zxkane) [inline examples/benchmark/README.md:12]: -> parameters

File: examples/benchmark/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+to run the program.
+
+2. Start performance test:
+Run below script to start performance test
```
Comment (zxkane) [inline examples/benchmark/README.md:16]: consolidate it as `benchmark` in doc and code

File: examples/benchmark/cleanup-ecs-performance-test.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
```
Comment (zxkane) [inline examples/benchmark/cleanup-ecs-performance-test.sh:2]: add bash options like `set -eo pipefail`

File: examples/benchmark/README.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+We have developed a tool for performance tests, which utilizes ECS Fargate to continuously send test data to Clickstream. The number of ECS Service tasks can be increased or decreased to adjust the RPS (Requests per Second).
+
+Follow below step to start your Clickstream performance test
+1. Add configuration file:
+
+Please download your `amplifyconfiguration.json` file from your web console of the solution, then copy it to the root
+folder which at the same level as README file. then you can execute the following command to create and send your sample
+data.
+
```
Comment (zxkane) [inline examples/benchmark/README.md:11]: add prerequesisities, for example,

- Python 3.xx
- Docker engine 
- configure `awscli` with proper privileges
- ...

**Review by zxkane — CHANGES_REQUESTED** [review]

File: examples/benchmark/Dockerfile (hunk 3 above)
Comment (dengmingtong) [reply examples/benchmark/Dockerfile:12]: I create and remove requirements.txt file in create-ecs-performance-test.sh.

File: examples/benchmark/Dockerfile (hunk 5 above)
Comment (dengmingtong) [reply examples/benchmark/Dockerfile:15]: Fixed

File: examples/benchmark/README.md (hunk 7 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:3]: fixed

File: examples/benchmark/README.md (hunk 8 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:12]: fixed

File: examples/benchmark/README.md (hunk 11 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:11]: Added

File: examples/benchmark/README.md (hunk 6 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:1]: fixed

File: examples/benchmark/README.md (hunk 9 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:16]: fixed

File: examples/benchmark/cleanup-ecs-performance-test.sh (hunk 10 above)
Comment (dengmingtong) [reply examples/benchmark/cleanup-ecs-performance-test.sh:2]: added

File: examples/benchmark/Dockerfile (hunk 4 above)
Comment (dengmingtong) [reply examples/benchmark/Dockerfile]: Refactored

File: examples/benchmark/Dockerfile (hunk 2 above)
Comment (dengmingtong) [reply examples/benchmark/Dockerfile]: Refactored

**Comment by aws-solution-clickstream-analytics** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|8.50|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/18/2024, 11:03:27 (UTC+0)_

**Comment by aws-solution-clickstream-analytics** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|8.50|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/18/2024, 11:28:41 (UTC+0)_

File: examples/benchmark/README.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
+## Clickstream benchmark tool
+
+We have developed a tool for benchmark tests, which utilizes ECS Fargate to continuously send mock data to Clickstream. The number of ECS Service tasks can be increased or decreased to adjust the RPS (Requests per Second).
+
+### Prerequesisities
+- Make sure you have an AWS account
+- Configure [credential of aws cli][configure-aws-cli]
```
Comment (zxkane) [inline examples/benchmark/README.md:7]: Is it a broken link after rendering the markdown?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: examples/benchmark/README.md (hunk 12 above)
Comment (dengmingtong) [reply examples/benchmark/README.md:7]: Yes, thank you to find this.
Fixed

**Comment by aws-solution-clickstream-analytics** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|8.50|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/21/2024, 12:54:54 (UTC+0)_

**Comment by aws-solution-clickstream-analytics** [comment]
### SonarQube Quality Gate Result
- **Result**: :white_check_mark: OK
- Triggered by @dengmingtong on `pull_request`

| Metric | Status | Value | Error Threshold |
|:------:|:------:|:-----:|:---------------:|
|Reliability rating|:white_check_mark: OK|1|> 1|
|Security rating|:white_check_mark: OK|1|> 1|
|Sqale rating|:white_check_mark: OK|1|> 1|
|Coverage|:white_check_mark: OK|83.70|< 80|
|Duplicated lines density|:white_check_mark: OK|8.50|> 30|
|Blocker violations|:white_check_mark: OK|0|> 0|
|Bugs|:white_check_mark: OK|0|> 0|
|Code smells|:white_check_mark: OK|9|> 40|
|Critical violations|:white_check_mark: OK|0|> 0|
|Major violations|:white_check_mark: OK|0|> 0|
|Vulnerabilities|:white_check_mark: OK|0|> 0|

[View on SonarQube](http://sonarqube:9000/dashboard?id=pr-1057)
###### _updated: 7/21/2024, 12:58:13 (UTC+0)_

**Review by zxkane — APPROVED** [review]

**Pull request merged**